File: easy/1496_path_crossing/solution.py
class Solution:
    def isPathCrossing(self, path: str) -> bool:
        x = y = 0
        visited = set()

        for direction in path:
            visited.add((x, y))

            if direction == "N":
                y += 1
            elif direction == "E":
                x += 1
            elif direction == "S":
                y -= 1
            elif direction == "W":
                x -= 1

            if (x, y) in visited:
                return True

        return False

File: easy/1496_path_crossing/solution.cpp
using namespace std;

#include <string>
#include <unordered_set>

class Solution {
public:
  bool isPathCrossing(string path) {
    int x = 0, y = 0;
    unordered_set<int> visited = {0};

    for (const char &direction : path) {
      switch (direction) {
      case 'N':
        ++y;
        break;
      case 'E':
        ++x;
        break;
      case 'S':
        --y;
        break;
      case 'W':
        --x;
        break;
      }

      const int hash = x * 67 + y;

      if (visited.find(hash) != visited.end())
        return true;

      visited.insert(hash);
    }

    return false;
  }
};
